tracer/src/Datadog.Trace/AppSec/Waf/ReturnTypes.Managed/InitializationResult.cs

```diff
@@ -0,0 +1,92 @@
+// <copyright file="InitializationResult.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using Datadog.Trace.AppSec.Waf.NativeBindings;
+using Datadog.Trace.Vendors.Newtonsoft.Json;
+
+namespace Datadog.Trace.AppSec.Waf.ReturnTypesManaged
+{
+    internal class InitializationResult
+    {
+        public InitializationResult(IntPtr? ruleHandle, ushort failedToLoadRules, ushort loadedRules, string ruleFileVersion, IReadOnlyDictionary<string, string[]> errors, bool unusableRuleFile = false)
+        {
+            HasErrors = errors.Count > 0;
+            Errors = errors;
+            ErrorMessage = HasErrors ? JsonConvert.SerializeObject(errors) : string.Empty;
+            FailedToLoadRules = failedToLoadRules;
+            LoadedRules = loadedRules;
+            RuleFileVersion = ruleFileVersion;
+            RuleHandle = ruleHandle;
+            UnusableRuleFile = unusableRuleFile;
+        }
+
+        internal IntPtr? RuleHandle { get; }
+
+        internal ushort FailedToLoadRules { get; }
+
+        /// <summary>
+        /// Gets the number of rules successfully loaded
+        /// </summary>
+        internal ushort LoadedRules { get; }
+
+        internal IReadOnlyDictionary<string, string[]> Errors { get; }
+
+        internal string ErrorMessage { get; }
+
+        internal bool HasErrors { get; }
+
+        internal bool UnusableRuleFile { get; }
+
+        internal string RuleFileVersion { get; }
+
+        internal string File { get; set; }
+
+        internal bool FileNotFound { get; set; }
+
+        internal bool ExceptionDecodingFile { get; set; }
+
+        internal static InitializationResult FromUnusableRuleFile() => new(null, 0, 0, string.Empty, new Dictionary<string, string[]>(), true);
+
+        internal static InitializationResult From(DdwafRuleSetInfoStruct ddwaRuleSetInfoStruct, IntPtr? ruleHandle)
+        {
+            var ddwafObjectStruct = ddwaRuleSetInfoStruct.Errors;
+            var errors = Decode(ddwafObjectStruct);
+            var ruleFileVersion = Marshal.PtrToStringAnsi(ddwaRuleSetInfoStruct.Version);
+            return new(ruleHandle, ddwaRuleSetInfoStruct.Failed, ddwaRuleSetInfoStruct.Loaded, ruleFileVersion, errors);
+        }
+
+        private static IReadOnlyDictionary<string, string[]> Decode(DdwafObjectStruct ddwafObjectStruct)
+        {
+            var nbEntriesStart = (int)ddwafObjectStruct.NbEntries;
+            var errorsDic = new Dictionary<string, string[]>(nbEntriesStart);
+            if (nbEntriesStart > 0)
+            {
+                var structSize = Marshal.SizeOf(typeof(DdwafObjectStruct));
+                for (var i = 0; i < nbEntriesStart; i++)
```

Comment:
The this loop, you make assumptions about the type of `DdwafObjectStruct` you receive, without checking what the actual type of the object is. It could be good to check the type of the object and print a warning if it's not as expected.

Reply:
true it should be a map , fixed in 50c1359ab57b5cc2a9503358084816b10d950f65